Adjust the dwl build for my own purposes.
--- a/Makefile
+++ b/Makefile
@@ -17,9 +17,13 @@ clean:
 install: dwl
 	install -Dm755 dwl $(PREFIX)/bin/dwl
 	install -Dm644 dwl.1 $(MANDIR)/man1/dwl.1
+	install -D dwl-session $(PREFIX)/bin/dwl-session
+	install -D dwl.desktop /usr/share/wayland-sessions/dwl.desktop
 
 uninstall:
 	rm -f $(PREFIX)/bin/dwl $(MANDIR)/man1/dwl.1
+	rm -f $(PREFIX)/bin/dwl-session
+	rm -f /usr/share/wayland-sessions/dwl.desktop
 
 .PHONY: all clean install uninstall
 
@@ -57,7 +61,7 @@ idle-protocol.c:
 idle-protocol.o: idle-protocol.h
 
 config.h: | config.def.h
-	cp config.def.h $@
+	cp -n config.def.h $@
 
 dwl.o: config.mk config.h client.h xdg-shell-protocol.h wlr-layer-shell-unstable-v1-protocol.h idle-protocol.h
 
